### ISSUE TITLE
Speeding up Uploads

### ISSUE THREAD

**Issue by Donmurry** [Duke-Translational-Bioinformatics/duke-data-service#1167]
Background:
As we started to look at the approach of batching chunks, we were looking at the sample project that has been used in the past (“handover project,” 1528 files, 75 dir, 36MB).  Each of those files only contains one chunk, so batching chunks really wouldn’t help the upload speed of this project.  This caused us to start digging into upload endpoints and the Postgres database to see where we can improve response times.

Problem Discovery:
As a first pass, D.Mann has a small script that hits the four DDS endpoints that are necessary to upload/create a file and loops 300 times. This creates enough traffic, over a long enough period of time, to show up in NewRelic. From there, we can investigate the component parts of the requests and identify slow areas. The script was run several times in `dukeds-analytics` and twice in `dukeds` (production). We've identified that a significant amount of time is spent on postgres(pg) calls. The pg calls that tend to take the most amount of time are interacting with the `job_transactions` and `audits` tables.

API-side addressing/testing problems:
1. We **upgraded our production version of pg** from 9.4 to 9.5.12.  Heroku currently offers pg versions 9.5, 9.6, and 10.0. It is not clear how upgrading will affect performance, but it needs to be done regardless.
2. Using our analytics instance, we can test the affects of **archiving the `job_transactions`** table on performance. This table has over 11 million rows in it currently. 
3. We can also test the affect of **archiving the `audits` table** using our analytics server. This table is queried throughout the application, so a more strategic solution will be needed in order to deploy audit archiving. But, that's not needed for the sake of testing performance.

Client-side questions:
1. During the upload process is the old projects/:id/children being called?  Would switching to projects/:id/files make a difference in upload performance?
2. Do you have any feedback/logs or further testing that could help us target what the client is experiencing?

We have to look at a more comprehensive approach to speeding up uploads, infrastructure as well as code.  As our anecdotal hypothesis of batching chunks won’t help in scenarios when each file only contains one chunk, and as we’ve discovered problems with Postgres, we’ll truly need to find out where there are problem areas.  @dleehr @johnbradley any help you can provide is appreciated.

**Comment by dleehr** [comment]
> Each of those files only contains one chunk, so batching chunks really wouldn’t help the upload speed of this project.

True, but the files could be batched together. This is a very common use case for the cores (e.g. upload this set of N files where n = 100, 200, 500, 1000, 2000).

I believe in the current implementation, to upload a 1-chunk file, the client makes 4 subsequent requests (3 to the service and 1 to the object store). Each request takes time. We parallelize them, but it adds up. 

On the download side, cutting down that number by an order of magnitude greatly sped things up, and I'm sure it would for uploads too.

> During the upload process is the old projects/:id/children being called? 

Yes, but the use case we benchmarked was uploading to a new, empty project. So there wouldn't be any children yet and this wouldn't contribute to the time.

When uploading to an existing project, the children endpoint is called to get metadata for comparing.

> Would switching to projects/:id/files make a difference in upload performance?

Not for the case we're benchmarking, but we do plan to use this batch endpoint to speed up that initial diff.

> Do you have any feedback/logs or further testing that could help us target what the client is experiencing?

Nothing we haven't already provided.

**Comment by johnbradley** [comment]
For single chunk files it would be nice to be able to request a batch of upload urls.
For each file right now we do:
- **POST** /api/v1/projects/{project_id}/uploads - _to create an upload_
- **PUT** /api/v1/uploads/{id}/chunks - _to get an upload url_
- upload the file to swift
- **PUT** /api/v1/uploads/{id}/complete - _make upload ready to become a file_
- **POST** /api/v1/files - _create file from upload_ or **PUT** /api/v1/files/{id} _to update a file_

One possibility that, might help for the small files case, would be to:
- create a batch of upload URLs for multiple files (combines creating upload and getting url)
- upload the file to swift (recreating urls if timeouts or errors occur)
- send batch complete request (combines completing uploads and creating files)

For large files it would need logic to only create the upload once and complete the upload once.

**Comment by johnbradley** [comment]
Possible payload to illustrate creating multiple file upload urls:
```
[
  {name: "file1", size:4000, chunk_size:10000,  // used for POST /projects/{project_id}/uploads
    hash: { value: "ablksjfd", algorithm: "md5"} // used for PUT /uploads/{id}/chunks
  },
  {name: "file2", size:5000, chunk_size:10000, // used for POST /projects/{project_id}/uploads
     hash: { value: "ablksjfd", algorithm: "md5"} // used for PUT /uploads/{id}/chunks
 }
]
```
Returns
```
[
  {name: "file1", upload_id: '<someid>', host: 'swift...', url: '/asldfkj...' },
  {name: "file2", upload_id: '<someid2>', host: 'swift...', url: '/asldfkj...' }
]
```

Possible payload to illustrate completing multiple file uploads (complete upload create files):
```
[ //used for POST /files
  {upload: {id: '<someid>'}, parent: {kind: 'dds-project', id:'13091340'}}, 
  {upload: {id: '<someid>'}, parent: {kind: 'dds-project', id:'13091340'}},  
]
```

We might need to be able to pass a unique id with each file so we could upload multiple files with the same name (in different directories for example). An endpoint to batch create folders might also help as well. Thanks.

**Comment by dmlond** [comment]
We really appreciate your input, and we definitely have put some thought into ways we could support fewer calls to these endpoints.  As we stated earlier, we first must solve some systemic issues that affect not only the time required to create objects, but the time required to query/serialize them in GET responses. During this process we will be establishing a baseline of the time required to do our current upload workflow process, e.g.

POST uploads (u)
PUT chunk(c)
POST to storage(s)
PUT complete (χ)
POST files || PUT file{id} (fe)

Once we solve these systemic issues, we will not only have sped up most calls to some extent without changing the API, but we will have a good set of baseline measurements of how long things take.

As we understand it, currently the total time to upload N files 
(Fte) =~ (N(u + cte + χ + fe))/Fp
Where
cte = c+s/cp
Fp = number of parallel processes which can manage these 4 steps for a given File.
cp = number of parallel processes which can manage the call to create the chunk and post the file to its storage.

We can then approach ideas on creating new endpoints or modifying our existing endpoints to reduce the number of calls required to upload files, and systematically test these ideas out against this baseline. 

For example, if we were to modify the files endpoint to support a request payload JSON with File attributes + upload + N Chunks, and return a JSON which contained the File, Upload, and Chunk Ids, plus the Chunk temp Urls, this would change the workflow to:
POST files || PUT file/{id} (fp)
POST to storage (s)
PUT complete (χ)

With Time to upload all files:
Ftp =~ (F(fp + ctp + χ))/Fp
Since ctp = s/cp, it should follow that ctp < cte
If fp < fe, then Ftp < Fte
We would then develop analytics tests to repeatedly run simulations of fp to determine and compare it to the baseline fe. If it looks promising, we could develop this endpoint.
 
We could model a batch complete endpoint and determine if Ftp < Fte in a similar manner.

One thing we are concerned about with regard to any REST endpoint that takes multiple File objects (e.g. sending a request to create [N File objects with their upload/chunks/hashes]) is that this will inevitably lead to request timeouts due to the time required to create and serialize the response for these objects, unless we implement it with strict limits of N. We are considering the investigation of websockets which could allow us to more fully utilize our existing background job infrastructure to allow clients to take advantage of it as well, e.g. a client would:
Connect
spawn thread1( send N requests to create N Files (with upload/chunks/hashes) )
spawn thread2( handle response from socket of File response with chunk temporary urls and spawn jobs/threads to handle it)
wait for threads to finish, possibly call complete(N) (unless it makes more sense to complete in the jobs/threads spawned by thread2)

In this case, the critical number would be the ΔTf (time between request sent in thread1 to response recieved by thread2). If ΔTf < fe + u, then this would be an improvement.
